### PR TITLE
Fix pinning in BaseOverlappedAsyncResult.CompletionCallback

### PR DESCRIPTION
**Pull request by stephentoub**
Some overloads of PostCompletion rely on the pinning configured by SetUnmanagedStructures, which means the call to it needs to come before the call to ReleaseUmanagedStructures, or else it could end up relying on the addresses of unpinned objects.

Fixes https://github.com/dotnet/corefx/issues/16953
cc: @geoffkizer, @cipop, @Priya91

### CODE REVIEW

**Comment by stephentoub** [comment]
(Corresponding desktop code: https://referencesource.microsoft.com/#System/net/System/Net/Sockets/_BaseOverlappedAsyncResult.cs,382)

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs [hunk 1]
```diff
@@ -133,10 +133,10 @@ private static unsafe void CompletionPortCallback(uint errorCode, uint numBytes,
         // Called either synchronously from SocketPal async routines or asynchronously via CompletionPortCallback above. 
         private void CompletionCallback(int numBytes, SocketError socketError)
         {
-            ReleaseUnmanagedStructures();
-
             ErrorCode = (int)socketError;
-            InvokeCallback(PostCompletion(numBytes));
+            object result = PostCompletion(numBytes);
+            ReleaseUnmanagedStructures();
```
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Add a comment so we don't accidentally break this again?

**Comment by geoffkizer** [comment]
Overall LGTM

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Windows.cs]: Added

**Pull request merged**